### PR TITLE
build(lint): disallow node process api, favor Toolkit ChildProcess api

### PR DESCRIPTION
**Pull request by Hweinstock**
## Problem
contributors may use the low-level node `child_process` library, which lacks the instrumentation and robustness of the toolkit `ChildProcess` utility. 

Example: https://github.com/aws/aws-toolkit-vscode/pull/5925#discussion_r1841277911
## Solution
- add the node `child_process` library as a "banned import" with a message pointing to our `ChildProcess` utility. 
- Migrate simple cases to use our `ChildProcess`, mark more complex ones as exceptions. 

## Future Work
- To migrate the existing cases marked as exceptions, we will need to add functionality to our `ChildProcess` api, such as enforcing a `maxBufferSize` on the output.
- Additionally, some of the current uses of `child_process` are tied to implementation details in the `child_process` library that make it hard to change without a larger-scale refactor. These cases were marked as exceptions. 

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -199,6 +199,11 @@ module.exports = {
                         name: 'fs',
                         message: 'Avoid node:fs and use shared/fs/fs.ts when possible.',
                     },
+                    {
+                        name: 'child_process',
+                        message:
+                            'Avoid child_process, use ChildProcess wrapper from `packages/core/src/shared/utilities/processUtils.ts` instead.',
```
Comment (justinmk3) [inline .eslintrc.js:205]:
```suggestion
                            'Avoid child_process, use ChildProcess from `shared/utilities/processUtils.ts` instead.',
```

File: packages/core/scripts/test/launchTestUtilities.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import * as proc from 'child_process'
+import * as proc from 'child_process' // eslint-disable-line no-restricted-imports
```
Comment (justinmk3) [inline packages/core/scripts/test/launchTestUtilities.ts:6]: 👍 in build-scripts we generally want to stick to plain old node APIs.

File: packages/core/src/shared/utilities/processUtils.ts [hunk 3]
```diff
@@ -100,6 +100,13 @@ export class ChildProcess {
         // TODO: allow caller to use the various loggers instead of just the single one
         this.#log = baseOptions.logging !== 'no' ? logger.getLogger() : logger.getNullLogger()
     }
+    public static async exec(
```
Comment (justinmk3) [inline packages/core/src/shared/utilities/processUtils.ts:103]: can we call this `run`? should avoid different names for the same thing.

File: packages/core/src/shared/sam/debugger/goSamDebug.ts [hunk 4]
```diff
@@ -187,7 +189,9 @@ async function makeInstallScript(debuggerPath: string, isWindows: boolean): Prom
 
     // It's fine if we can't get the latest Delve version, the Toolkit will use the last built one instead
     try {
-        const goPath: string = JSON.parse(execFileSync('go', ['env', '-json']).toString()).GOPATH
+        const goPath: string = JSON.parse(
+            (await ChildProcess.exec('go', ['env', '-json'], { collect: true })).stdout
+        ).GOPATH
```
Comment (justinmk3) [inline packages/core/src/shared/sam/debugger/goSamDebug.ts:194]: Use a local var is a bit more terse and readable, at the cost of having to claim a name.
But generally, invoking a process almost always should involve checking an exit code (though that's out of scope for this PR), so inlining `exec()` or `run()` is unusual and likely discouraged. .

```suggestion
        const r = await ChildProcess.exec('go', ['env', '-json'], { collect: true })
        const goPath: string = JSON.parse(r.stdout).GOPATH
```

**Review by justinmk3 — APPROVED** [review]

**Comment by justinmk3** [comment]
This is good to have in place. We don't need to bother migrating any of the child_process cases for Q features, since most of those will use Flare.

File: packages/core/src/shared/utilities/processUtils.ts (hunk 3 above)
Comment (Hweinstock) [reply packages/core/src/shared/utilities/processUtils.ts:103]: Original intention was to mirror `child_process` `exec`, but I agree the consistency is more important.

File: packages/core/src/shared/sam/debugger/goSamDebug.ts (hunk 4 above)
Comment (Hweinstock) [reply packages/core/src/shared/sam/debugger/goSamDebug.ts:194]: Ideally checking the exit code would be part of some higher-level API provided by ChildProcess right?

**Review by justinmk3 — APPROVED** [review]

File: packages/core/src/shared/sam/debugger/goSamDebug.ts (hunk 4 above)
Comment (justinmk3) [reply packages/core/src/shared/sam/debugger/goSamDebug.ts]: > Ideally checking the exit code would be part of some higher-level API provided by ChildProcess right?

Failure handling in some way needs to be done by the caller, typically

**Pull request merged**